phylogeny/prototypeSelection.py

```diff
@@ -0,0 +1,106 @@
+"""
+Prototype selection
+
+Given a set of n elements (S) and their pairwise distances in terms of a
+dissimilarity matrix (DM) and a given number (k << n), find the sub-set (s)
+consisting of exactly k elements, such that s best represents the full set S.
+
+Here, we define "best represents" as maximizing the sum of distances between
+all points, i.e. sum {over a,b in S} DM[a,b]. This is our objective function.
+
+This problem is known to be NP-hard [1], thus we need to resort to heuristics.
+In the future, this module will implement several heuristics, whose quality can
+be measured by the objective function for each problem instance, since there is
+no global winner.
+For completeness, the exact but exponential algorithm is implemented, too.
+  "prototypeSelection_exhaustive"
+
+[1] Gamez, J. Esteban, François Modave, and Olga Kosheleva.
+    "Selecting the most representative sample is NP-hard:
+     Need for expert (fuzzy) knowledge."
+    Fuzzy Systems, 2008. FUZZ-IEEE 2008.
+"""
+
+from skbio.stats.distance import DistanceMatrix
+import itertools
+import numpy as np
+import scipy as sp
+
+
+def distanceSum(elements, dm: DistanceMatrix):
+    '''Compute the sum of pairwise distances for the given elements according to
+    the given dissimilarity matrix.
+
+    Parameters
+    ----------
+    elements: list or set
+        list or set of element IDs for which the sum of distances is computed
+    dm: skbio.stats.distance.DistanceMatrix
+        pairwise dissimilarity matrix.
+
+    Returns
+    -------
+    float:
+        the sum of all pairwise distances of dm for IDs in elements
+    '''
+
+    # some assertions
+    if type(dm) != DistanceMatrix:
+        raise TypeError('dm is not of type "DistanceMatrix"')
+    assert(len(elements) > 1)
+    assert(len(elements) <= len(dm.ids))
```

Comment:
Guido is working on compile time type checking but still in dev and not in
Python proper:

http://mypy-lang.org

Doing the annotations does hurt but is above and beyond what we do, for
instance, in scikit bio (although I could see a push to put them in since
that project is now py3 only).

It doesn't hurt. But I so recommend against explicit type checks except
perhaps in outer, user facing methods or high level API calls. It will
consume a large amount of runtime if they're done within a loop or nested
loop (generally aim to avoid those checks deep in the call stack)

On Sep 2, 2016 09:43, "Stefan Janssen" <notifications@github.com> wrote:

In phylogeny/prototypeSelection.py
<https://github.com/biocore/WGS-HGT/pull/32#discussion_r77376519>:

> +    elements: list or set
> +        list or set of element IDs for which the sum of distances is computed
> +    dm: skbio.stats.distance.DistanceMatrix
> +        pairwise dissimilarity matrix.
> +
> +    Returns
> +    -------
> +    float:
> +        the sum of all pairwise distances of dm for IDs in elements
> +    '''
> +
> +    # some assertions
> +    if type(dm) != DistanceMatrix:
> +        raise TypeError('dm is not of type "DistanceMatrix"')
> +    assert(len(elements) > 1)
> +    assert(len(elements) <= len(dm.ids))

so, what style should I use here? Personally, I would like to annotate
every type wherever possible. Is that sufficient and can I thus remove the
(maybe expensive) type checking or should it remain, because python (at the
moment) does not enforce types?

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/biocore/WGS-HGT/pull/32/files/f12a6fa86bd8fbfda95231d3b1a11d318a85014d#r77376519>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAc8slCK9azz-H1CIpVe3JFzXYlIbiQkks5qmFJHgaJpZM4JyK-v>
.